### PR TITLE
Make chalk-engine more UnwindSafe

### PR DESCRIPTION
**Pull request by jackh726**
cc. https://github.com/rust-lang/chalk/issues/260

I can't call this 100% UnwindSafe, so I'm not going to say this *closes* the linked issue. But I think with proper tests added to confirm, it might be.

### CODE REVIEW

File: chalk-engine/src/logic.rs [hunk 1]
```diff
@@ -77,11 +78,18 @@ impl<C: Context> Forest<C> {
         table: TableIndex,
         answer_index: AnswerIndex,
     ) -> RootSearchResult<CompleteAnswer<C>> {
-        assert!(self.stack.is_empty());
+        let stack = Stack::default();
```
Comment (nikomatsakis) [inline chalk-engine/src/logic.rs]: interesting

**Comment by jackh726** [comment]
@matklad and @flodiebold 

This might make it so that rust-analyzer doesn't have to reset chalk state on panic

**Comment by matklad** [comment]
>I can't call this 100% UnwindSafe, so I'm not going to say this closes the linked issue. But I think with proper tests added to confirm, it might be.

So this doesn't literally implement `UnwindSafe` trait? I think it might be good to add such an impl even if we are not 100% sure it is correct, and let rust-analyzer find the bugs in practice? We unwind a lot, so, if something somewhere asserts, we'll notice it :D

**Comment by jackh726** [comment]
If I get some time to add at least a couple tests (at least for regression in the future), then I'll also add `UnwindSafe` and close the related issue.

I would definitely like to see if everything is good for rust-analyzer without resetting state. I'm not sure if there are some corner cases that might be hit anywhere.

**Pull request merged**